# Lab | Data Aggregation and Filtering

In this challenge, we will continue to work with customer data from an insurance company. We will use the dataset called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv

### COMMENT: Hello! Sabina, i beleive we should be using the CLEAN CSV not this one.

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by first performing data cleaning, formatting, and structuring.

1. Create a new DataFrame that only includes customers who:
   - have a **low total_claim_amount** (e.g., below $1,000),
   - have a response "Yes" to the last marketing campaign.

2. Using the original Dataframe, analyze:
   - the average `monthly_premium` and/or customer lifetime value by `policy_type` and `gender` for customers who responded "Yes", and
   - compare these insights to `total_claim_amount` patterns, and discuss which segments appear most profitable or low-risk for the company.

3. Analyze the total number of customers who have policies in each state, and then filter the results to only include states where there are more than 500 customers.

4. Find the maximum, minimum, and median customer lifetime value by education level and gender. Write your conclusions.

## Bonus

5. The marketing team wants to analyze the number of policies sold by state and month. Present the data in a table where the months are arranged as columns and the states are arranged as rows.

6.  Display a new DataFrame that contains the number of policies sold by month, by state, for the top 3 states with the highest number of policies sold.

*Hint:*
- *To accomplish this, you will first need to group the data by state and month, then count the number of policies sold for each group. Afterwards, you will need to sort the data by the count of policies sold in descending order.*
- *Next, you will select the top 3 states with the highest number of policies sold.*
- *Finally, you will create a new DataFrame that contains the number of policies sold by month for each of the top 3 states.*

7. The marketing team wants to analyze the effect of different marketing channels on the customer response rate.

Hint: You can use melt to unpivot the data and create a table that shows the customer response rate (those who responded "Yes") by marketing channel.

External Resources for Data Filtering: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

In [20]:
# your code goes here

---
# Solved Lab | Data Aggregation and Filtering
---

In this challenge, we will continue to work with customer data from an insurance company. We will use the dataset called `marketing_customer_analysis.csv`, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. 

Our goal is to explore and analyze this data by first performing data cleaning (already cleaned it), formatting, and structuring.

---

#### Load and View Dataset

In [21]:
# import necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from IPython.display import display
import seaborn as sns
plt.style.use('ggplot')
pd.set_option('display.max_columns', 200)

# Load online_shoppers_intention.csv into a df_main DataFrame
df = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv")

# Display df DataFrame SHAPE
print("RAW Dataset:")
print(f"Dataset Shape: [{df.shape[0]}] ROWS and [{df.shape[1]}] COLUMNS\n")
display(df.head())

RAW Dataset:
Dataset Shape: [10910] ROWS and [27] COLUMNS



,unnamed:_0,customer,state,customer_lifetime_value,response,coverage,education,effective_to_date,employmentstatus,gender,income,location_code,marital_status,monthly_premium_auto,months_since_last_claim,months_since_policy_inception,number_of_open_complaints,number_of_policies,policy_type,policy,renew_offer_type,sales_channel,total_claim_amount,vehicle_class,vehicle_size,vehicle_type,month
0,0,DK49336,Arizona,4809.216960,No,Basic,College,2011-02-18,Employed,M,48029,Suburban,Married,61,7.000000,52,0.000000,9,Corporate Auto,Corporate L3,Offer3,Agent,292.800000,Four-Door Car,Medsize,A,2
1,1,KX64629,California,2228.525238,No,Basic,College,2011-01-18,Unemployed,F,0,Suburban,Single,64,3.000000,26,0.000000,1,Personal Auto,Personal L3,Offer4,Call Center,744.924331,Four-Door Car,Medsize,A,1
2,2,LZ68649,Washington,14947.917300,No,Basic,Bachelor,2011-02-10,Employed,M,22139,Suburban,Single,100,34.000000,31,0.000000,2,Personal Auto,Personal L3,Offer3,Call Center,480.000000,SUV,Medsize,A,2
3,3,XL78013,Oregon,22332.439460,Yes,Extended,College,2011-01-11,Employed,M,49078,Suburban,Single,97,10.000000,3,0.000000,2,Corporate Auto,Corporate L3,Offer2,Branch,484.013411,Four-Door Car,Medsize,A,1
4,4,QA50777,Oregon,9025.067525,No,Premium,Bachelor,2011-01-17,Medical Leave,F,23675,Suburban,Married,117,15.149071,31,0.384256,7,Personal Auto,Personal L2,Offer1,Branch,707.925645,Four-Door Car,Medsize,A,1


---
1. Create a new DataFrame that only includes customers who:
   - have a **low total_claim_amount** (e.g., below $1,000),
   - have a response "Yes" to the last marketing campaign.

In [22]:
# Filter the DataFrame that includes customers with total_claim_amount below $1,000 AND response "Yes"
df_filtered = df[(df['total_claim_amount'] < 1000) & (df['response'] == 'Yes')]

# Display the filtered DataFrame
display(df_filtered.head())

,unnamed:_0,customer,state,customer_lifetime_value,response,coverage,education,effective_to_date,employmentstatus,gender,income,location_code,marital_status,monthly_premium_auto,months_since_last_claim,months_since_policy_inception,number_of_open_complaints,number_of_policies,policy_type,policy,renew_offer_type,sales_channel,total_claim_amount,vehicle_class,vehicle_size,vehicle_type,month
3,3,XL78013,Oregon,22332.439460,Yes,Extended,College,2011-01-11,Employed,M,49078,Suburban,Single,97,10.0,3,0.0,2,Corporate Auto,Corporate L3,Offer2,Branch,484.013411,Four-Door Car,Medsize,A,1
8,8,FM55990,California,5989.773931,Yes,Premium,College,2011-01-19,Employed,M,66839,Suburban,Single,154,33.0,24,0.0,1,Personal Auto,Personal L1,Offer2,Branch,739.200000,Sports Car,Medsize,A,1
15,15,CW49887,California,4626.801093,Yes,Basic,Master,2011-01-16,Employed,F,79487,Suburban,Divorced,114,20.0,87,0.0,1,Special Auto,Special L1,Offer2,Branch,547.200000,SUV,Medsize,A,1
19,19,NJ54277,California,3746.751625,Yes,Extended,College,2011-02-26,Employed,F,41479,Rural,Married,94,14.0,38,1.0,1,Personal Auto,Personal L2,Offer2,Call Center,19.575683,Two-Door Car,Large,A,2
27,27,MQ68407,Oregon,4376.363592,Yes,Premium,Bachelor,2011-02-28,Employed,F,63774,Rural,Divorced,111,18.0,63,0.0,1,Personal Auto,Personal L3,Offer2,Agent,60.036683,Four-Door Car,Medsize,A,2


In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10910 entries, 0 to 10909
Data columns (total 27 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   unnamed:_0                     10910 non-null  int64  
 1   customer                       10910 non-null  object 
 2   state                          10910 non-null  object 
 3   customer_lifetime_value        10910 non-null  float64
 4   response                       10910 non-null  object 
 5   coverage                       10910 non-null  object 
 6   education                      10910 non-null  object 
 7   effective_to_date              10910 non-null  object 
 8   employmentstatus               10910 non-null  object 
 9   gender                         10910 non-null  object 
 10  income                         10910 non-null  int64  
 11  location_code                  10910 non-null  object 
 12  marital_status                 10910 non-null 

---
2. Using the original Dataframe, analyze:
   - the average `monthly_premium` and/or customer lifetime value by `policy_type` and `gender` for customers who responded "Yes", 
   
   and
   - compare these insights to `total_claim_amount` patterns, and discuss which segments appear most profitable or low-risk for the company.

In [24]:
print("average monthly premium auto:", df["monthly_premium_auto"].mean())
display(df["monthly_premium_auto"].describe().round(2))


print(70*"-")
print("average customer lifetime value by policy:")
print(70*"-")
display(df["customer_lifetime_value"].groupby(df["policy"]).mean().round(2))


print(70*"-")
print("average customer lifetime value by policy:")
print(70*"-")
display(df_filtered["customer_lifetime_value"].groupby(df_filtered["policy"]).mean().round(2))

average monthly premium auto: 93.19605866177818


count    10910.00
mean        93.20
std         34.44
min         61.00
25%         68.00
50%         83.00
75%        109.00
max        298.00
Name: monthly_premium_auto, dtype: float64

----------------------------------------------------------------------
average customer lifetime value by policy:
----------------------------------------------------------------------


policy
Corporate L1    8324.08
Corporate L2    7688.65
Corporate L3    7805.30
Personal L1     7947.27
Personal L2     8081.68
Personal L3     8016.19
Special L1      8023.63
Special L2      8515.41
Special L3      9255.36
Name: customer_lifetime_value, dtype: float64

----------------------------------------------------------------------
average customer lifetime value by policy:
----------------------------------------------------------------------


policy
Corporate L1    8389.67
Corporate L2    7010.11
Corporate L3    7706.72
Personal L1     7096.49
Personal L2     8078.55
Personal L3     7732.74
Special L1      6787.56
Special L2      6853.64
Special L3      8999.20
Name: customer_lifetime_value, dtype: float64

Grouping directly by both policy_type and gender for customers with response 'Yes'

In [25]:
print("average monthly premium auto:", df_filtered["monthly_premium_auto"].mean())
display(df_filtered["monthly_premium_auto"].describe().round(2))


print(70*"-")
print("average customer lifetime value by policy:")
print(70*"-")
display(df_filtered["customer_lifetime_value"].groupby(df_filtered["policy"]).mean().round(2))


print(70*"-")
print("average customer lifetime value by policy:")
print(70*"-")
display(df_filtered["customer_lifetime_value"].groupby(df_filtered["policy"]).mean().round(2))

average monthly premium auto: 88.7205146533238


count    1399.00
mean       88.72
std        22.64
min        61.00
25%        67.00
50%        84.00
75%       107.00
max       154.00
Name: monthly_premium_auto, dtype: float64

----------------------------------------------------------------------
average customer lifetime value by policy:
----------------------------------------------------------------------


policy
Corporate L1    8389.67
Corporate L2    7010.11
Corporate L3    7706.72
Personal L1     7096.49
Personal L2     8078.55
Personal L3     7732.74
Special L1      6787.56
Special L2      6853.64
Special L3      8999.20
Name: customer_lifetime_value, dtype: float64

----------------------------------------------------------------------
average customer lifetime value by policy:
----------------------------------------------------------------------


policy
Corporate L1    8389.67
Corporate L2    7010.11
Corporate L3    7706.72
Personal L1     7096.49
Personal L2     8078.55
Personal L3     7732.74
Special L1      6787.56
Special L2      6853.64
Special L3      8999.20
Name: customer_lifetime_value, dtype: float64

---
3. Analyze the total number of customers who have policies in each state, and then filter the results to only include states where there are more than 500 customers.

In [26]:
# number of customer grouped by state and policy
df_customers_states_policy = (
    df.groupby(["state", "policy"]).size().reset_index(name="customer_count")
)
display(df_customers_states_policy)
# total number of customers who have policies above 500 in each state
print("total number of customers who have policies above 500 in each state:")
df_customer_count_by_state = df_customers_states_policy[df_customers_states_policy["customer_count"] > 500].groupby("state")["customer_count"].sum()
display(df_customer_count_by_state)

,state,policy,customer_count
0,Arizona,Corporate L1,62
1,Arizona,Corporate L2,119
2,Arizona,Corporate L3,200
3,Arizona,Personal L1,293
4,Arizona,Personal L2,434
5,Arizona,Personal L3,742
6,Arizona,Special L1,25
7,Arizona,Special L2,35
8,Arizona,Special L3,27
9,California,Corporate L1,199


total number of customers who have policies above 500 in each state:


state
Arizona        742
California    3058
Oregon        1775
Name: customer_count, dtype: int64

In [27]:
df_states = (
    df
    .groupby("state")
    .size()
    .reset_index(name="customer_count")
)

display(df_states)

,state,customer_count
0,Arizona,1937
1,California,4183
2,Nevada,993
3,Oregon,2909
4,Washington,888


**Data Insight:**
Customer Distribution by State (>500 customers)
| State | Total Number of Customers|
|---|---|
|California | 3_058|
|Oregon|1_775|
|Arizona | 742 |

**🔎 Business Insight:**

- California has the largest customer base by a wide margin, representing the primary market.
- Oregon holds a strong secondary position with a substantial number of customers.
- Arizona, while smaller, still exceeds the 500-customer threshold and represents a stable market segment.
- All listed states meet the filtering condition of more than 500 customers, meaning no states were excluded after applying the threshold.

**📊 Business Insight:**
- California should be prioritized for retention and upselling strategies due to its high customer concentration.
- Oregon presents solid growth and cross-selling opportunities.
- Arizona may offer targeted expansion potential, as increasing share here could significantly impact overall growth.

---
4. Find the maximum, minimum, and median customer lifetime value by education level and gender. Write your conclusions.


In [28]:
# display customer_lifetime_value by education level and gender
print(70*"-")
print("customer_lifetime_value by education level and gender:")
# mean customer_lifetime_value by education level and gender
display(df.groupby(["education", "gender"])["customer_lifetime_value"].mean().round(2))
print(70*"-")
# maximum customer_lifetime_value by education level and gender
display(df.groupby(["education", "gender"])["customer_lifetime_value"].max().round(2))
print(70*"-")
# minimum customer_lifetime_value by education level and gender
display(df.groupby(["education", "gender"])["customer_lifetime_value"].min().round(2))

----------------------------------------------------------------------
customer_lifetime_value by education level and gender:


education             gender
Bachelor              F         7874.27
                      M         7703.60
College               F         7748.82
                      M         8052.46
Doctor                F         7328.51
                      M         7415.33
High School or Below  F         8675.22
                      M         8149.69
Master                F         8157.05
                      M         8168.83
Name: customer_lifetime_value, dtype: float64

----------------------------------------------------------------------


education             gender
Bachelor              F         73225.96
                      M         67907.27
College               F         61850.19
                      M         61134.68
Doctor                F         44856.11
                      M         32677.34
High School or Below  F         55277.45
                      M         83325.38
Master                F         51016.07
                      M         50568.26
Name: customer_lifetime_value, dtype: float64

----------------------------------------------------------------------


education             gender
Bachelor              F         1904.00
                      M         1898.01
College               F         1898.68
                      M         1918.12
Doctor                F         2395.57
                      M         2267.60
High School or Below  F         2144.92
                      M         1940.98
Master                F         2417.78
                      M         2272.31
Name: customer_lifetime_value, dtype: float64

---

**Data Insight:**
**1. Higher Education Is Associated With Higher CLV**
- Customers with Master’s and Doctor degrees have the highest lifetime values.
    - Master: F (2417.78), M (2272.31)
    - Doctor: F (2395.57), M (2267.60)
- This suggests that more educated customers tend to generate more long-term revenue.

**2. Bachelor and College Levels Have Lower CLV:**
- These groups may represent mid-value segments compared to postgraduate customers.
    - Bachelor and College degree holders show the lowest CLV overall.
    - Bachelor: F (1904.00), M (1898.01)
    - College: F (1898.68), M (1918.12)

**3. Gender Differences Vary by Education Level**
- Females have higher CLV than males in most education categories:
    - Bachelor,
    - Doctor,
    - High School or Below,
    - Master.
- Exception: In the College category, males (1918.12) slightly exceed females (1898.68).

**4. High School or Below Shows an Interesting Pattern**
- Females (2144.92) have noticeably higher CLV than males (1940.98).
- This gap is larger than in most other education categories.

**5. Highest and Lowest Segments**
- Highest CLV overall: Female customers with a Master’s degree (2417.78).
- Lowest CLV overall: Male customers with a Bachelor’s degree (1898.01).

**Business Insight:**
- Postgraduate customers (Master’s and Doctorate) are the most valuable segment.
- Female customers tend to have slightly higher lifetime value across most education levels.
- Targeted retention and premium offerings may be especially effective for highly educated female customers.


---
## Bonus

5. The marketing team wants to analyze the number of policies sold by state and month. Present the data in a table where the months are arranged as columns and the states are arranged as rows.

6.  Display a new DataFrame that contains the number of policies sold by month, by state, for the top 3 states with the highest number of policies sold.

*Hint:*
- *To accomplish this, you will first need to group the data by state and month, then count the number of policies sold for each group. Afterwards, you will need to sort the data by the count of policies sold in descending order.*
- *Next, you will select the top 3 states with the highest number of policies sold.*
- *Finally, you will create a new DataFrame that contains the number of policies sold by month for each of the top 3 states.*

7. The marketing team wants to analyze the effect of different marketing channels on the customer response rate.

Hint: You can use melt to unpivot the data and create a table that shows the customer response rate (those who responded "Yes") by marketing channel.

External Resources for Data Filtering: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

In [29]:
# your code goes here